compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -355,6 +365,41 @@ fn irrefutable_let_pattern(tcx: TyCtxt<'_>, span: Span, id: HirId, source: hir::
     });
 }
 
+fn check_if_let_guard<'p, 'tcx>(
+    cx: &mut MatchCheckCtxt<'p, 'tcx>,
+    pat: &'p super::Pat<'tcx>,
+    pat_id: HirId,
+) {
+    let mut seen = Matrix::empty();
+    let v = PatStack::from_pattern(pat);
+    match is_useful(cx, &seen, &v, LeaveOutWitness, pat_id, false, true) {
+        NotUseful => unreachable_pattern(cx.tcx, pat.span, pat_id, None),
+        Useful(unreachables) => {
+            let mut unreachables: Vec<_> = unreachables.into_iter().flatten().collect();
+            // Emit lints in the order in which they occur in the file.
+            unreachables.sort_unstable();
+            for span in unreachables {
+                unreachable_pattern(cx.tcx, span, pat_id, None);
+            }
+        }
+        UsefulWithWitness(_) => bug!(),
+    }
+    seen.push(v);
+
+    let wildcard_pat = super::Pat::wildcard_from_ty(pat.ty);
+    let v = PatStack::from_pattern(&wildcard_pat);
+    match is_useful(cx, &seen, &v, LeaveOutWitness, pat_id, false, true) {
+        NotUseful => irrefutable_let_pattern(
+            cx.tcx,
+            pat.span,
+            pat_id,
+            hir::MatchSource::IfLetDesugar { contains_else_clause: true },
+        ),
+        Useful(_) => {}
+        UsefulWithWitness(_) => bug!(),
+    }
+}
+
```

Comment:
Soryy, I did a bit of a refactor in https://github.com/rust-lang/rust/pull/78995. I think you need something like this now:

```rust
fn check_if_let_guard<'p, 'tcx>(
    cx: &mut MatchCheckCtxt<'p, 'tcx>,
    pat: &'p super::Pat<'tcx>,
    pat_id: HirId,
) {
    let arms = [MatchArm { pat, hir_id: pat_id, has_guard: false }];
    let report = compute_match_usefulness(&cx, &arms, pat_id, pat.ty);

    match &report.arm_usefulness[0] {
        NotUseful => unreachable_pattern(cx.tcx, pat.span, pat_id, None),
		Useful(unreachables) if unreachables.is_empty() => {}
        Useful(unreachables) => {
            let mut unreachables: Vec<_> = unreachables.into_iter().flatten().collect();
            // Emit lints in the order in which they occur in the file.
            unreachables.sort_unstable();
            for span in unreachables {
                unreachable_pattern(cx.tcx, span, pat_id, None);
            }
        }
        UsefulWithWitness(_) => bug!(),
    }

    if report.non_exhaustiveness_witnesses.is_empty() {
        // The match is exhaustive, i.e. the if let pattern is irrefutable.
        irrefutable_let_pattern(
            cx.tcx,
            pat.span,
            pat_id,
            hir::MatchSource::IfLetDesugar { contains_else_clause: true },
        )
    }
}
```

The printing of unreachable subpatterns is duplicate with what's happening in `report_arm_reachability`. Is there a way you could deduplicate it? Maybe adding a `MatchSource::IfLetGuardDesugar`?